### ISSUE TITLE
project-select: update sort of projects in list

### ISSUE THREAD

**Issue by spadgett** [openshift/origin-web-catalog#390]
The list of projects in the select-project dropdown should:

1. Show the 5 most recent, and then have a divider with everything else sorted alphabetically...
2. ...by display name, case insensitive

**Comment by dtaylor113** [comment]
Hi, here is what I have so far:

![image](https://user-images.githubusercontent.com/12733153/29573979-467eabe6-872e-11e7-97bc-d098752d43c6.png)
![image](https://user-images.githubusercontent.com/12733153/29573990-4b3c4cce-872e-11e7-8183-81d7a2058983.png)

* The top 5 are most recently viewed/created
* The remaining projects (below divider) are sorted by displayName, case insensitive

@spadgett, @serenamarie125, do you think we need a divider between "Create Project" and the top 5 most recently added/created?

**Comment by spadgett** [comment]
If you're using `ui-select` groups, you can give each section a title and it will add a divider automatically.

https://github.com/angular-ui/ui-select/wiki/ui-select-choices

**Comment by spadgett** [comment]
I think you should able able to do this with just a simple `group-by` function and no other changes

**Comment by dtaylor113** [comment]
Ok, thanks @spadgett,  do we want a divider between "Create Projects" and rest?

**Comment by spadgett** [comment]
I think we probably need it because it won't fit in the other categories. Do you mind adding a screenshot when you have it working?

**Comment by dtaylor113** [comment]
Sure, np.  So for the ```groupBy``` it takes a string or a function, which do you prefer?  Should I add a ```project.groupBy``` property to the projects objects, or a ```RecentlyViewedProjectsService.isRecentlyViewedProject(item.metadata.uid)``` to determine the grouping?

**Comment by dtaylor113** [comment]
> I think you should able able to do this with just a simple group-by function and no other changes

I updated ```RecentlyViewedProjectsService.orderByMostRecentlyViewed(projects, secondarySort)``` to allow the secondary sort to be specified.  By default its sorted by mostRecentlyViewed, then by creationDate (this is what's used in Project Summary), ```RecentlyViewedProjectsService.orderByMostRecentlyViewed(projects, 'displayName')``` will sort first by mostRecentlyViewed followed by displayName (case insensitive).

**Comment by spadgett** [comment]
> So for the groupBy it takes a string or a function, which do you prefer?

Probably a function.

> Should I add a project.groupBy property to the projects objects

Don't add properties to the project objects because they'll be cached and reused from now on. You shouldn't mutate them.

> RecentlyViewedProjectsService.isRecentlyViewedProject(item.metadata.uid) to determine the grouping?

Yeah, I think you'll need this.

**Comment by dtaylor113** [comment]
Ok, thanks, that what I thought

**Comment by dtaylor113** [comment]
Here's with ```groupBy```:

![image](https://user-images.githubusercontent.com/12733153/29577167-839370de-8738-11e7-9b90-8b6a6b51b01c.png)
![image](https://user-images.githubusercontent.com/12733153/29577174-894db78c-8738-11e7-8cd0-f53d31b64322.png)

```
public groupChoicesBy = (item: any) => {
    if (this.RecentlyViewedProjectsService.isRecentlyViewed(item.metadata.uid)) {
      return "recentlyViewed";
    } else if (item.metadata.annotations["openshift.io/display-name"] === "Create Project") {
      return "create";
    } else {
      return "project";
    }
  }
```
We should decide on what the groupBy labels should be.

**Comment by spadgett** [comment]
I'd say "Recent" and "All Project" for the last two. I might try `""` for create if it works

**Comment by dtaylor113** [comment]
> I'd say "Recent" and "All Project" for the last two. I might try "" for create if it works

Hmm...it's not really 'All Projects' is remaining projects minus the recentlyViewed, how about "Other Projects"

**Comment by dtaylor113** [comment]
![image](https://user-images.githubusercontent.com/12733153/29577636-262a262a-873a-11e7-849b-1b012da825a5.png)
![image](https://user-images.githubusercontent.com/12733153/29577641-2a5c4246-873a-11e7-86f2-fef153dbfdb8.png)

**Comment by spadgett** [comment]
@openshift/team-ux-review Opinion?

I do like the shorter "Recent" or perhaps "Recent Projects" for the first group

**Comment by ncameronbritt** [comment]
It seems a little redundant to me to include "project" in all of the labels when everything in the menu is a project (other than the "Create Project" action).  

What about "Recently Viewed"?

**Comment by cshinn** [comment]
Looks good! I agree with @ncameronbritt that "Recently Viewed" seems like a good label.

**Comment by dtaylor113** [comment]
Ok, sounds good.  What about the 'Other Projects' label?  Should that just be 'Others'?

**Comment by beanh66** [comment]
Yes @dtaylor113 I agree with the previous comments as well and I would think "Recently Viewed" and "Others" would work fine. I do agree with you that the "Create Project" kinda blends in though. Wondering if that should be Bold or Blue Link font. @jennyhaines any thoughts?

**Comment by cshinn** [comment]
I personally prefer "Other Projects". Others feels vague and sort of ominous

**Comment by beanh66** [comment]
haha good point @cshinn, works for me!

**Comment by ncameronbritt** [comment]
Does "other projects" need a label at all?

To @beanh66 's point, we use the blue link text in the "View all projects" link in the project selection menu. 
<img width="505" alt="screen shot 2017-08-22 at 1 28 27 pm" src="https://user-images.githubusercontent.com/8974593/29578685-2a291778-873e-11e7-9fb7-155185abab6d.png">

**Comment by dtaylor113** [comment]
The 'View all projects' in project summary is a link.   The 'Other Projects' is a label/category in a dropdown list.

**Comment by dtaylor113** [comment]
Here I was able to just use a space for the "other projects" catelgory label:

![image](https://user-images.githubusercontent.com/12733153/29578945-29455cf8-873f-11e7-93a9-e788884bd032.png)
![image](https://user-images.githubusercontent.com/12733153/29578951-2e093f7a-873f-11e7-8ded-f60ebd0b2770.png)

I personally prefer "Other Projects" to no label.

**Comment by beanh66** [comment]
@dtaylor113 I think he was maybe voting in favor of the "Create Project" being a blue link and I would agree. But @ncameronbritt do I have that correct?

**Comment by ncameronbritt** [comment]
That's right, @beanh66. But I see how my comment was confusing. Sorry!

**Comment by dtaylor113** [comment]
> @dtaylor113 I think he was maybe voting in favor of the "Create Project" being a blue link and I would agree.

Ok, but it's not a link, it's a dropdown selection, seems strange to me to color it to look like a link, but then again, I'm not a designer :-)

**Comment by spadgett** [comment]
I tend to agree. I'd expect it to navigate me away if it's styled like a link.

**Comment by dtaylor113** [comment]
How about making it italic to indicate it's a special choice?

**Comment by ncameronbritt** [comment]
Good point @dtaylor113 and @spadgett. Guess all that option does is add three additional fields. That's different from the project list example I gave. 

I think what you've got is good.

**Comment by beanh66** [comment]
Agreed, good points all around. If anything I would bold the create project or leave it as is.

**Comment by dtaylor113** [comment]
Fixed via #404
